### PR TITLE
feat(api): add auto-pagination and server-side aggregation

### LINKED ISSUE #80: feat: Add auto-pagination and server-side aggregation for deals and invoices
## 🎯 Goal

Eliminate manual pagination and LLM-side aggregation by adding auto-pagination and server-side summarization for list endpoints.

**Definition of Done:**

- [ ] Generic `fetchAllPages()` method handles all pagination automatically
- [ ] New aggregation tools return pre-computed summaries to LLM
- [ ] LLM receives structured summaries, not raw deal/invoice arrays

### Why This Matters

Currently, list endpoints (`freee_get_deals`, `freee_get_invoices`) return at most 100 records per call with manual `offset`/`limit` pagination. This forces the LLM to:
1. Make multiple sequential API calls to fetch all records
2. Aggregate numbers across pages (error-prone, token-expensive)
3. Waste context window on raw JSON instead of insights

The freee API rate limit is 3,600 req/hr. Fetching 1,000 deals requires 10 API calls just for data retrieval.

### Current State

- List tools return raw paginated results (max 100 per page)
- LLM must manually manage offset/limit and aggregate across pages
- No server-side summarization capabilities

### Desired State

- Server handles all pagination internally
- New tools return aggregated financial summaries
- LLM focuses on interpreting results, not fetching data

### User Stories

- As an LLM user, I want to get a full summary of deals for a period so that I can analyze financial trends without manual pagination
- As an LLM user, I want invoice payment status summaries so that I can quickly identify cash flow issues

### Functional Requirements

1. **Auto-pagination in FreeeClient**
   - Add a generic `fetchAllPages()` method that handles offset/limit pagination automatically:
     ```typescript
     private async fetchAllPages<T>(
       endpoint: string,
       params: Record<string, unknown>,
       dataKey: string,
       maxPages?: number
     ): Promise<T[]> {
       const allResults: T[] = [];
       let offset = 0;
       const limit = 100;
       while (true) {
         const response = await this.client.get(endpoint, { params: { ...params, offset, limit } });
         const items = response.data[dataKey] || [];
         allResults.push(...items);
         if (items.length < limit || (maxPages && offset / limit >= maxPages - 1)) break;
         offset += limit;
       }
       return allResults;
     }
     ```

2. **`freee_search_deals`** — Fetch all deals matching criteria, return aggregated summary
   - Input: companyId, dateRange, partnerId, accountItemId (all optional filters)
   - Output: `{ totalCount, totalIncome, totalExpense, byPartner: [{name, income, expense}], byMonth: [{month, income, expense}], byAccountItem: [{name, total}] }`
   - Internally: auto-paginate all matching deals, compute aggregations server-side

3. **`freee_summarize_invoices`** — Fetch all invoices, return payment status summary
   - Input: companyId, dateRange, partnerId (optional)
   - Output: `{ totalCount, totalAmount, unpaidAmount, overdueCount, byStatus: [{status, count, amount}], byPartner: [{name, count, amount, unpaid}] }`

4. **Enhance existing list tools**
   - Add `fetchAll: boolean` parameter to existing `freee_get_deals` and `freee_get_invoices`
   - When true, auto-paginate and return all results (with configurable `maxRecords` safety limit, default 1000)

### Technical Constraints

- **Technology**: TypeScript, Zod schemas for MCP tool registration
- **Performance**: Safety limit prevents unbounded pagination (default 1000 records max)
- **Rate limit**: Log warning if approaching 3,600 req/hr limit
- **Compatibility**: Must work with existing MCP SDK 1.x tool registration pattern

### Functional

- [ ] `fetchAllPages()` generic method added to FreeeClient
- [ ] `freee_search_deals` tool registered with Zod schema
- [ ] `freee_summarize_invoices` tool registered with Zod schema
- [ ] Aggregation computed server-side (not in LLM)
- [ ] Rate limit awareness: log warning if approaching limit

### Testing

- [ ] Unit tests for aggregation logic (deals summary, invoice summary)
- [ ] Unit tests for pagination edge cases (empty results, exactly 100 results, partial page)
- [ ] `npm run build && npm test` pass

### Affected Files

- `src/api/freeeClient.ts` — Add `fetchAllPages()` method and aggregation methods
- `src/index.ts` — Register new tools (`freee_search_deals`, `freee_summarize_invoices`)
- `src/schemas.ts` — Add Zod schemas for new tools
- `src/types/freee.ts` — Add TypeScript interfaces for aggregated responses
- `src/__tests__/api/freeeClient.test.ts` — Unit tests

### Existing Patterns

- Follow `server.tool()` registration pattern in `src/index.ts`
- Follow FreeeClient method pattern in `src/api/freeeClient.ts`

### Key Design Decisions

- Server returns **aggregated summaries** to LLM, not raw deal arrays
- LLM focuses on **interpreting results** (trends, anomalies, recommendations)
- Safety limit prevents unbounded pagination (default 1000 records max)
- Respects 3,600 req/hr rate limit

## 🚧 Risks & Mitigation

| Risk | Impact | Probability | Mitigation |
|------|--------|-------------|------------|
| Rate limit hit during large pagination | High | Medium | maxPages safety limit, rate limit logging |
| Memory pressure from large datasets | Medium | Low | maxRecords cap (default 1000), streaming if needed |
| freee API timeout on slow queries | Medium | Low | Per-request timeout, retry logic |

## 🔗 Related Work

- Related to #63 (structured logging will help debug pagination)

### 🟡 Recommended

- [ ] Test with Claude Desktop to verify tool descriptions guide LLM effectively
- [ ] Measure API call reduction in typical analysis sessions
- [ ] Verify aggregation accuracy against manual calculations

### ⚠️ Manual Verification

- [ ] Test with large datasets (500+ deals) in staging
- [ ] Monitor API rate limit consumption

## 🏷️ Metadata

**Priority**: High
**Size**: L
**Type**: Feature
**Labels**: enhancement, priority:high

---

🤖 AI-Optimized Issue
Generated with [Claude Code](https://claude.com/claude-code)

### PR DESCRIPTION
**Pull request by knishioka**
## Summary

Closes #80

- **Auto-pagination**: Generic `fetchAllPages()` method with offset/limit pagination (100 per page, 1000 max safety cap) for all list endpoints
- **`freee_search_deals` tool**: Fetches all matching deals and returns server-side aggregated summaries by partner, month, and account item
- **`freee_summarize_invoices` tool**: Fetches all matching invoices with payment status breakdown, overdue detection, and per-partner analysis

## Changes

| File | Change |
|------|--------|
| `src/types/freee.ts` | Added 7 aggregation interfaces (`DealAggregation`, `InvoiceSummaryAggregation`, etc.) |
| `src/constants.ts` | Added `PAGINATION_LIMIT` (100) and `MAX_AUTO_PAGINATION_RECORDS` (1000) |
| `src/schemas.ts` | Added `SearchDealsSchema` and `SummarizeInvoicesSchema` with `maxRecords` field |
| `src/api/freeeClient.ts` | Added `fetchAllPages()`, `searchDeals()`, `summarizeInvoices()`, and private aggregation helpers |
| `src/index.ts` | Registered 2 new MCP tools (`freee_search_deals`, `freee_summarize_invoices`) |
| `src/__tests__/api/pagination.test.ts` | New: 18 tests for pagination and aggregation |
| `src/__tests__/index.test.ts` | Updated tool/schema counts from 18 → 20 |

## Design Decisions

- **Safety limit**: 1000 records max per auto-pagination call to respect freee's 3,600 req/hour rate limit
- **Server-side aggregation**: Summaries computed before sending to LLM, reducing token usage significantly
- **Existing patterns**: Uses the established `registerTool` wrapper and `FreeeClient` method patterns

## Test plan

- [x] `fetchAllPages` pagination: single page, multi-page, empty, boundary (exactly 100), maxRecords limit
- [x] `searchDeals` aggregation: by partner, month, account item; empty deals; deals without partners; filter passthrough
- [x] `summarizeInvoices`: by status/partner; overdue detection; empty invoices; empty payment_status; filter passthrough
- [x] All 251 tests pass (8 suites)
- [x] TypeScript typecheck passes
- [x] ESLint passes (warnings only)
- [x] Build compiles successfully

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/api/freeeClient.ts [hunk 1]
```diff
@@ -617,5 +626,292 @@ export class FreeeClient {
...
+  }
+
+  private aggregateInvoices(
+    invoices: FreeeInvoice[],
+  ): InvoiceSummaryAggregation {
+    const sorted = [...invoices].sort((a, b) =>
+      a.issue_date.localeCompare(b.issue_date),
+    );
+
+    const today = new Date().toISOString().split('T')[0];
```
[the remark this answers is not in the document]
Comment (knishioka) [reply src/api/freeeClient.ts]: Fixed in commit 3449f82. Same fix as the gemini thread above — replaced `new Date().toISOString().split('T')[0]` with `Intl.DateTimeFormat` using `timeZone: 'Asia/Tokyo'` for correct JST date calculation.
